xls/modules/dbe/lz4_decoder.x

```diff
@@ -0,0 +1,618 @@
+    i_ram_hb_rd_resp: chan<ram::ReadResp<SYM_WIDTH>> in;
+
+    init {
+        State<SYM_WIDTH, PTR_WIDTH, CNT_WIDTH> {
+            wr_ptr: uN[PTR_WIDTH]:0,
+            rd_ptr: uN[PTR_WIDTH]:0,
+            rd_ctr_rem: uN[CNT_WIDTH]:0,
+            err: false,
+            hb_all_valid: false,
+        }
+    }
+
+    config (
+        i_encoded: chan<Token<SYM_WIDTH, PTR_WIDTH, CNT_WIDTH>> in,
+        o_symb: chan<PlainData<SYM_WIDTH>> out,
+        // RAM
+        o_ram_hb_wr_req: chan<ram::WriteReq<PTR_WIDTH, SYM_WIDTH, 1>> out,
+        i_ram_hb_wr_resp: chan<ram::WriteResp> in,
+        o_ram_hb_rd_req: chan<ram::ReadReq<PTR_WIDTH, 1>> out,
+        i_ram_hb_rd_resp: chan<ram::ReadResp<SYM_WIDTH>> in,
+    ) {
+        (
+            i_encoded, o_symb,
+            o_ram_hb_wr_req, i_ram_hb_wr_resp,
+            o_ram_hb_rd_req, i_ram_hb_rd_resp,
+        )
+    }
+
+    next (tok: token, state: State<SYM_WIDTH, PTR_WIDTH, CNT_WIDTH>) {
+        let is_copying = state.rd_ctr_rem != uN[CNT_WIDTH]:0;
+        let do_recv = !is_copying;
+
+        // Receive token
+        let (tok, rx) = recv_if(tok, i_encoded, do_recv,
+            dbe::zero_token<SYM_WIDTH, PTR_WIDTH, CNT_WIDTH>());
+
+        let rx_lt = do_recv && rx.kind == TokenKind::LT;
+        let rx_cp = do_recv && rx.kind == TokenKind::CP;
+        let rx_mark = do_recv && rx.kind == TokenKind::MK;
+        let rx_end = rx_mark && rx.mark == Mark::END;
+        let rx_error = rx_mark && dbe::is_error(rx.mark);
+        let rx_reset = rx_mark && rx.mark == Mark::RESET;
+        let rx_unexpected_mark = rx_mark && !(rx_end || rx_reset || rx_error);
+
+        // Are we going to output a literal symbol or a symbol from HB?
+        let emit_sym_lit = rx_lt;
+        let emit_sym_from_hb = rx_cp || is_copying;
+
+        // Reset, set or increment read pointer
+        let rd_ptr = if rx_reset {
+            uN[PTR_WIDTH]:0
+        } else if rx_cp {
+            state.wr_ptr - rx.cp_off - uN[PTR_WIDTH]:1
+        } else if is_copying {
+            state.rd_ptr + uN[PTR_WIDTH]:1
+        } else {
+            state.rd_ptr
+        };
+
+        // Check whether rd_ptr points to an unwritten HB cell
+        let rd_ptr_valid = (rd_ptr < state.wr_ptr) || state.hb_all_valid;
+
+        // HB READ - read the symbol from history buffer
+        let do_hb_read = emit_sym_from_hb && rd_ptr_valid;
+        let tok = send_if(tok, o_ram_hb_rd_req, do_hb_read,
+            ram::ReadWordReq<u32:1>(rd_ptr));
+        let (tok, hb_rd) = recv_if(tok, i_ram_hb_rd_resp, do_hb_read,
+            RamZeroReadResp<SYM_WIDTH>());
+
+        // Generate ERROR_INVAL_CP when we should've read from HB but
+        // we couldn't because CP points to an invalid location
+        let err_inval_cp = emit_sym_from_hb && !do_hb_read;
+
+        // Send decoded data symbol or Mark
+        let zero_data = dbe::zero_plain_data<SYM_WIDTH>();
+        let (data_valid, data) = if rx_reset {
```

Comment:
rx_rest is more about clearing the state to handle a new block, right?  If so, consider changing the name to reflect that.

Reply:
In general it's an equivalent to a complete state reset: it both prepares the proc to handle a new block of data, as well as it clears a sticky error state that the proc (and other procs in the chain) may be in.

Reply:
It seems the common case is to handle the new block of data, no?  I'd suggest naming it that rather than the less common case of resetting the error state.

Reply:
For now I've kept the "reset" name, as the idea of "RESET" token was to reset a full network of processing blocks (e.g. LZ4 block encoder or anything else that can be there besides this single core), and it just happened that in the encoder this function coincides with the operation that has to be done when you want to start processing of a new independent block. I can rename it to "restart" or something like that, or even add a new marker code, but that would be redundant (in a sense that if you're starting to encode an independent block of data, you don't want to preserve any previous state, so why not just completely reset the circuit?).

Reply:
I'd say the issue here is that the reset signal is usually special -- and may be considered a multi-cycle signal.

It sounds like this signal here is expected to be a mission-mode signal that should be timed as such -- so it's more a personal preference?